OD-527-code-integrity

amossCivicA opened the pull request:
modified ckan/update-requirements.sh and ckan/Dockerfile.dev to support hashes

woodt reviewed (changes requested):
I tested this out, building the images, updating the dependencies, rebuilding the ckan image, tweaking a dependency, updating the dependencies again, and then finally rebuilding the ckan image and testing it.  All seemed to work.

Could you add/update the "Requirement management" section of the README to reflect your changes?

Also, there is a minor issue: in the `requirements-noh.txt` file, the git URLs contain tags or branch names, not a "resolved" SHA.  This is not really a problem with this PR -- I suspect this was broken in the existing code.  You can see what I'm talking about if you compare the corresponding lines from `requirements-noh.txt` with `requirements-freeze.txt` (after running `make freeze-dependencies`)

I'm not sure what could be done about this.  It's not ideal, since we're not quite as "pinned" as we should be for these dependencies.

amossCivicA commented on ckan/Dockerfile.dev:
```diff
@@ -55,7 +55,9 @@ RUN curl -o ${SRC_DIR}/get-pip.py https://bootstrap.pypa.io/pip/2.7/get-pip.py &
 COPY requirements-dev.txt ${APP_DIR}
-RUN pip install -r ${APP_DIR}/requirements-dev.txt
+COPY requirements-noh.txt ${APP_DIR}
+RUN "pip" "install" "-r" "${APP_DIR}/requirements-dev.txt"
```
I saw this done while searching for syntax, but it has been changed.

amossCivicA commented on ckan/update-requirements.sh:
```diff
@@ -8,12 +8,29 @@ cd requirements
+# Generate the required format
+for i in `cat ../ckan/requirements-noh.tmp`; do PKGNM=`echo $i | awk -F: '{print $1}'`; PKGDTL=`echo $i | awk -F/ '{print $4"/"$5}'`;echo "${PREFIX}${PKGDTL}#egg=${PKGNM}" >> ../ckan/requirements-noh.txt ; done
+# Add the final touch
+#sed -i 's/^git/\-e git/g' ../ckan/requirements-noh.txt
```
removed this line that was used during testing

amossCivicA commented:
@woodt All my tests were successful.  Ready for review

woodt commented on ckan/update-requirements.sh:
```diff
@@ -8,12 +8,18 @@ cd requirements
+grep -v https ../ckan/requirements.tmp > ../ckan/requirements.txt
+grep -v https ../ckan/requirements-dev.tmp > ../ckan/requirements-dev.txt
+# Now extract the packages that require editable (non-hashed) source from the hashable packages
+grep https ../ckan/requirements-freeze.txt | grep -v httplib > ../ckan/requirements-noh.txt
```
Two questions about this line: (1) why exclude httplib?  and (2) there is a bit of a chicken and egg problem here, since requirements-freeze.txt is also a derived file.  If I understand this code correctly, you'll be pulling an old requirements line from requirements-freeze.txt into requirements-now.txt.  Consider what would happen if we were trying to update the version of something that ends up in requirements-noh.txt.  If you could demonstrate/document how this would work, that's great.  But I'm not sure that it would.

amossCivicA replied:
I have modified the update-requirements script to retrieve the correct commit hash for the appropriate editable source from the poetry.lock file.  I believe this should correct all the issues.

woodt replied:
So, not an unreasonable idea to pull the hash out of the poetry lock file! I do have a few concerns:
(1) Not clear to me that the poetry lock file has a guaranteed schema.  Do you know if the Poetry project has said that they have committed to that file format/contents?
(2) If we assume that is OK we read the file, the code in the last commit seems quite brittle (e.g., if Poetry adds a couple of keys to individual packages, I think the grep context size (the "-A3" arg) might need to be changed.  I'd feel *much* better if we were using code that was TOML aware to parse this file.  Probably lots of options, but I think the Python 'toml' library would work.

amossCivicA commented:
Modified the process to call the github api to retrieve the correct hash for the editable source.
Everything was working except for one package that was failing.  
It took at litte time to discover that some results were being truncated.  
Added "?per_page=1000" to rectify that.

woodt commented:
I pulled this branch and ran a `make update-dependencies` and then a `make build`.  The update-dependencies succeeded (yay!) but the build failed.  The egg names in the requirements-noh.txt file are incorrect (e.g., the CKAN egg should be named "ckan", not "ckan/ckan")

It is important that you test with `make build` after you make any changes to the generated requirements.
